File: pb/src/main/java/org/sqlbuilder/pb/objects/Rel.java
package org.sqlbuilder.pb.objects;

import org.sqlbuilder.common.*;
import org.sqlbuilder.pb.PbNormalizer;

import java.util.Comparator;

public class Rel implements HasId, Insertable, Comparable<Rel>
{
	private static final Comparator<Rel> COMPARATOR = Comparator //
			.comparing(Rel::getExample) //
			.thenComparing(Rel::getText) //
			.thenComparing(Rel::getF, Comparator.nullsFirst(Comparator.naturalOrder()));

	public static final SetCollector<Rel> COLLECTOR = new SetCollector<>(COMPARATOR);

	private final String text;

	private final Func f;

	private final Example example;

	// C O N S T R U C T O R

	public static Rel make(final Example example, final String text, final Func f)
	{
		var r = new Rel(example, text, f);
		COLLECTOR.add(r);
		return r;
	}

	private Rel(final Example example, final String text, final Func f)
	{
		this.text = PbNormalizer.normalize(text);
		this.f = f;
		this.example = example;
	}

	// A C C E S S

	public Example getExample()
	{
		return example;
	}

	public String getText()
	{
		return text;
	}

	public Func getF()
	{
		return f;
	}

	@RequiresIdFrom(type = Rel.class)
	@Override
	public Integer getIntId()
	{
		return COLLECTOR.get(this);
	}

	// O R D E R

	@Override
	public int compareTo(@NotNull final Rel that)
	{
		return COMPARATOR.compare(this, that);
	}

	@Override
	public String dataRow()
	{
		// (relid),rel,func,exampleid
		return String.format("'%s',%s,%s", //
				Utils.escape(text), //
				Utils.nullable(f, Func::getSqlId), //
				example.getSqlId() //
		);
	}

	@Override
	public String toString()
	{
		return String.format("rel %s[%s][%s]", text, example, f);
	}
}

File: pb/src/main/java/org/sqlbuilder/pb/objects/Rel.kt
package org.sqlbuilder.pb.objects

import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.HasId
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.SetCollector
import org.sqlbuilder.common.Utils
import org.sqlbuilder.pb.PbNormalizer

class Rel private constructor(val example: Example, text: String) : HasId, Insertable, Comparable<Rel> {

    val text: String = PbNormalizer.normalize(text)

    // N I D

    @RequiresIdFrom(type = Rel::class)
    override fun getIntId(): Int {
        return COLLECTOR.invoke(this)
    }

    // O R D E R

    override fun compareTo(that: Rel): Int {
        return COMPARATOR.compare(this, that)
    }

    override fun dataRow(): String {
        return "'${Utils.escape(text)}',${example.getSqlId()}"
    }

    override fun toString(): String {
        return "rel $text[$example]"
    }

    companion object {

        private val COMPARATOR: Comparator<Rel> = Comparator
            .comparing<Rel, Example> { it.example }
            .thenComparing<String> { it.text }

        @JvmField
        val COLLECTOR = SetCollector<Rel>(COMPARATOR)

        fun make(example: Example, text: String): Rel {
            val r = Rel(example, text)
            COLLECTOR.add(r)
            return r
        }
    }
}
